keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1197 @@
+ResourceClaimParameters type) which components such as a scheduler or
+autoscaler can evaluate against the NodeResourceSlices of candidate nodes,
+without knowing exactly what is being requested.
+
+When a workload starts, an on-node driver is responsible for making the
+requested device(s) active in the Pod, according to the resource choices made
+by the scheduler.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+### Non-Goals
+
+- Support semantic class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Scheduling performance is expected to become better compared to using the
+  PodSchedulingContext. However, this is not the reason for this KEP.
+
+
+## Proposal
+
+### Publishing node resources
+
+The kubelet publishes NodeResourceSlices with content provided by the different
+drivers running on the node. Access control through the node authorizer ensures
+that one kubelet is not allowed to create or modify NodeResourceSlices
+belonging to other nodes.
+
+NodeResourceSlices are published separately for each driver, using a version of
+the API that matches what that driver is using. This implies that kubelet, DRA
+driver, and apiserver must all support that same API version. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+Embedded inside the NodeResourceSlices is the list of resources managed by the
+driver. Typically there is a single object per driver, but this could also get
+spread out across multiple if a single object would be too large.
+
+Here is an example where the driver provides one discrete card which has
+a certain amount of compute units and RAM that can be reserved:
+
+```
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+nodeName: worker-1
+driverName: cards.dra.example.com
+resources:
+- name: CardA
```

Comment:
Let's say that a node has 2 cards like this. Would the NRSlice contain two elements with roughly the same contents?

Reply:
Details on this normalization are still being worked out, but our current thinking is that we have at least some data normalized so we don't explode the size of the data.

Metadata about a card can be captured once and then published once to the NodeResourceSlice, perhaps, and then individual instances refer back to it. So you might end up with a very rich "here is what my card looks like" and then a list of 8 cards that are just a name/UID and a reference back to that list. We are also talking about capturing the *status* of actual allocations somewhere - this would have to refer to the name/UID.